public/app/plugins/datasource/loki/datasource.ts

```diff
@@ -176,16 +185,36 @@ export class LokiDatasource extends DataSourceApi<LokiQuery, LokiOptions> {
       .map(target => this.prepareLiveTarget(target, options));
 
     for (const liveTarget of liveTargets) {
-      const subscription = webSocket(liveTarget.url)
+      liveTarget.subscription = webSocket(liveTarget.url)
         .pipe(
-          map((results: any[]) => {
-            const delta = this.processResult(results, liveTarget);
+          map((response: any) => {
+            const { refId } = liveTarget;
+
+            // What should we do with:
+            // response.dropped_entries?
+
+            const streams: LokiLogsStream[] = response.streams;
+            if (streams) {
+              for (const stream of streams) {
+                const buffer = getBufferedDataForLiveTarget(stream.labels, liveTarget);
```

Comment:
I was really scratching my head and thinking this is bug that you do not do anything with the buffer here afterwards. It took me a while to realise `getBufferedDataForLiveTarget` modifies the `liveTargets`. I know in grand scale that is kind of nitpicking but that kind of API is really bad. Either return something new or modify the parameter but doing both is highly confusing.